### PR TITLE
bash autocompletion for argocd

### PR DESCRIPTION
**Pull request by edwinpjacques**
# bash autocompletion for argocd

Resolve issue #1798 request to implement bash autocompletion for argocd.

Doing this required updating the version of cobra and flag dependencies.  Made them the same as argo.

## testing

To test it, this is what I did (OS X):
```
make cli
sudo cp -R ./dist/argocd `which argocd`
argocd completion bash >/usr/local/etc/bash_completion.d/argocd
. /usr/local/etc/bash_completion.d/argocd
```

## example

I tried to create reasonable completions for everything.  Here's just an example

```
Edwins-MacBook-Pro:argo-cd edwin$ argocd 
account     app         cluster     completion  context     login       logout      proj        relogin     repo        version
Edwins-MacBook-Pro:argo-cd edwin$ argocd proj 
add-destination           allow-namespace-resource  deny-cluster-resource     get                       remove-source
add-source                create                    deny-namespace-resource   list                      role
allow-cluster-resource    delete                    edit                      remove-destination        set
Edwins-MacBook-Pro:argo-cd edwin$ argocd proj role 
add-policy     create         create-token   delete         delete-token   get            list           remove-policy  
Edwins-MacBook-Pro:argo-cd edwin$ argocd proj role create default 
superman     wonderwoman  
Edwins-MacBook-Pro:argo-cd edwin$ argocd proj role create default ironman
Role 'ironman' created
```

### CODE REVIEW

File: cmd/argocd/commands/completion.go
Comment (alexec) [inline cmd/argocd/commands/completion.go]: Can I please ask - is this generated?

File: cmd/argocd/commands/cluster.go
Comment (alexec) [inline cmd/argocd/commands/cluster.go]: Nice.

File: cmd/argocd/commands/completion.go
Comment (alexec) [inline cmd/argocd/commands/completion.go]: Great example.

**Comment by alexec** [comment]
I think the CI failure is unrelated to the changes.

File: cmd/argocd/commands/completion.go
Comment (alexec) [reply cmd/argocd/commands/completion.go]: https://github.com/spf13/cobra/blob/master/bash_completions.md

Comment (alexec) [reply cmd/argocd/commands/completion.go]: I think this can be auto-generated, but it is not?

Comment (edwinpjacques) [reply cmd/argocd/commands/completion.go]: Yeah, I tried that and I got an error and could not figure out how to recover.

```
Edwins-MacBook-Pro:argocd edwin$ cobra add completion
Error: open /Users/edwin/go/src/github.com/argoproj/argo-cd/cmd/argocd/cmd/completion.go: no such file or directory
Edwins-MacBook-Pro:argocd edwin$ cobra --help
Cobra is a CLI library for Go that empowers applications.
This application is a tool to generate the needed files
to quickly create a Cobra application.

Usage:
  cobra [command]

Available Commands:
  add         Add a command to a Cobra Application
  help        Help about any command
  init        Initialize a Cobra Application

Flags:
  -a, --author string    author name for copyright attribution (default "YOUR NAME")
      --config string    config file (default is $HOME/.cobra.yaml)
  -h, --help             help for cobra
  -l, --license string   name of license for the project
      --viper            use Viper for configuration (default true)

Use "cobra [command] --help" for more information about a command.
```

I'm going to guess this is because the argocd is leveraging older cobra conventions, though I did not do the code archaeology to confirm.

So, what I did was copy the completion.go from the Argo Project. The custom completion bash code still has to be hand-written.  Cobra will make all the static completions work automatically so it's a lot less work.

**Comment by edwinpjacques** [comment]
> I think the CI failure is unrelated to the changes.

I was trying to figure out what was failing.  I even set up my own circle-ci.  What's wrong?  Is it because of some other changes?

**Comment by edwinpjacques** [comment]
@alexec should I suppose I leave it to you to mark the conversations as resolved.  Thanks!

File: cmd/argocd/commands/app.go
Comment (alexec) [inline cmd/argocd/commands/app.go]: can we use a default here?

File: cmd/argocd/commands/completion.go
Comment (alexec) [reply cmd/argocd/commands/completion.go]: ah - I get it. I'd be strongly in favor of generating this, as I know that we'll always forget to fix this. I might try myself shortly.

File: cmd/argocd/commands/project.go
Comment (alexec) [inline cmd/argocd/commands/project.go]: should this default to `wide`?

**Comment by alexec** [comment]
If the error is "_get_comp_words_by_ref command not found ", see: 
https://github.com/jweslley/rails_completion/issues/2

File: cmd/argocd/commands/cluster.go
Comment (alexec) [inline cmd/argocd/commands/cluster.go]: should this be `name` rather than `server` (consistent with `kubectl`)? and `wide` by default?

**Review by alexec — APPROVED** [review]
I've downloaded this and played around and happily got it to work completing the built-in commands. However, I think you want to go one step beyond and have it auto-complete app names? The way this is done automatically is using `ValidArgs`, but that can only encode a static list of values. Obviously, app names change between installs.

I'd happily approve and merge this PR if I can be sure that it won't create a maintenance overhead. I'd do that happily without the auto-completion of app names.

However, when I generate the commands, the file is 2,398 lines long. The code in `completion.go` is only 200 lines long. The final file is 2577 lines long. Am I right in thinking they combine the two?

I'm going to work on that assumption and approve if you want to make some small changes.

**Review by alexec — CHANGES_REQUESTED** [review]
Couple of minor comments about default values.

File: cmd/argocd/commands/completion.go
Comment (edwinpjacques) [reply cmd/argocd/commands/completion.go]: Well, to be clear, the code that's generated is boilerplate.  It's essentially static.  The only thing that's custom are the little scripts to do the automated lookups of things like app history, app names, repos, projects, namespaces, and roles.  The only maintenance would be if the CLI by with the lookups are done is changed, which I think is unlikely.

File: cmd/argocd/commands/app.go
Comment (edwinpjacques) [reply cmd/argocd/commands/app.go]: Yes, it's good to be explicit.  Will look for other such cases in code I changed.

Comment (edwinpjacques) [reply cmd/argocd/commands/app.go]: For this particular argument, I just have the same default argument, which is blank.  I did not change it.  I think that's appropriate.  The user can lean on completion if they are using the command line interactively IMO.

File: cmd/argocd/commands/cluster.go
Comment (edwinpjacques) [reply cmd/argocd/commands/cluster.go]: The cluster list has a name column, but it's blank, and appears optional.  I needed something that would always work.  So I took the server column.  I can make the default output wide.

``` bash
Edwins-MacBook-Pro:argo-cd edwin$ argocd cluster list
SERVER                          NAME  STATUS      MESSAGE
https://kubernetes.default.svc        Successful  
```

File: cmd/argocd/commands/project.go
Comment (edwinpjacques) [reply cmd/argocd/commands/project.go]: Sure, I can take care of that as well.  In the logic, it will be wide anyway, but better to be explicit.

File: cmd/argocd/commands/completion.go
Comment (edwinpjacques) [reply cmd/argocd/commands/completion.go]: In other words, I basically did the same thing the auto-generator would do.  And unless we went to a new major version of cobra, I don't think the generator will ever have to be re-run.  So it should be no/low maintenance.

**Comment by edwinpjacques** [comment]
> I've downloaded this and played around and happily got it to work completing the built-in commands. However, I think you want to go one step beyond and have it auto-complete app names? The way this is done automatically is using `ValidArgs`, but that can only encode a static list of values. Obviously, app names change between installs.
> 
> I'd happily approve and merge this PR if I can be sure that it won't create a maintenance overhead. I'd do that happily without the auto-completion of app names.
> 
> However, when I generate the commands, the file is 2,398 lines long. The code in `completion.go` is only 200 lines long. The final file is 2577 lines long. Am I right in thinking they combine the two?
> 
> I'm going to work on that assumption and approve if you want to make some small changes.

The magical completion of app names is not hard-coded.  I implemented the dynamic lookup code in completion.go, in the function __argocd_list_apps:

``` bash
__argocd_list_apps() {
	local -a argocd_out
	if argocd_out=($(argocd app list --output name 2>/dev/null)); then
		COMPREPLY+=( $( compgen -W "${argocd_out[*]}" -- "$cur" ) )
	fi
}
```

Is it ok to keep that?

**Comment by edwinpjacques** [comment]
I basically made everything dynamic.  If there are new commands added, then the default (static) completion of fixed options will still be generated automatically.  However, the dynamic lookup of completions will not happen unless someone adds cases into the bashCompletionFunc.  It's ok if someone does not, it does not hurt or break anything, just less convenient for interactive command-line use.  So again there should be little/no maintenance.  If any concerns remain, please let me know, I will be sure to address them!

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
Merging!

**Pull request merged**
